### PR TITLE
Fix duplicate HTLC fail-back on stale force-close

### LINKED ISSUE #3570: Add test for duplicate HTLC fail-back
In 0.1.1 we fixed an issue where if someone relays two identical (hash and amount) HTLCs through us and then force-closes with a stale state we'll only fail-back one of them. We fixed it indirectly by just failing back everything at a timeout, which is fine, but we shouldn't wait to fail these HTLCs back. Further, we should add a test for this.

### PR DESCRIPTION
**Pull request by Anyitechs**
This PR addresses the duplicate HTLC fail-back on stale force-close that fail-back one of them and waits till timeout to fail-back everything. This change adds proper tracking of HTLCs by payment hash to ensure all identical HTLCs are failed back immediately when detected, rather than waiting for timeout.

Also introduces a functional test that demonstrates the scenario by creating two identical HTLCs, performing a stale force-close, and verifying both HTLCs are failed back correctly.

Fixes #3570

### CODE REVIEW

**Comment by Anyitechs** [comment]
This PR is still work in progress. Need some help with the test, I'm able to simulate the scenario described in the issue but the change introduced is triggered when the node restarts and I'm not sure how to simulate that (did a network disconnection and reconnection, but that doesn't seem to do the job).

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -15695,15 +15695,28 @@ where
...
-					for (channel_htlc_source, payment_hash) in channel.inflight_htlc_sources() {
-						let mut found_htlc = false;
-						for (monitor_htlc_source, _) in monitor.get_all_current_outbound_htlcs() {
-							if *channel_htlc_source == monitor_htlc_source {
-								found_htlc = true;
-								break;
-							}
-						}
-						if !found_htlc {
+					// Collecting all outbound HTLC sources into a HashSet allows for efficient,
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs]: No, the `ChannelMonitor` is the source of the issue here - the issue was about HTLCs being failed back from the on-chain logic in `ChannelMonitor` (which is responsible for HTLCs once we close a channel) so it should be fixed, not worked around here.

Comment (Anyitechs) [reply lightning/src/ln/channelmanager.rs]: Ah, thank you for pointing that out.

File: lightning/src/chain/channelmonitor.rs [hunk 2]
```diff
@@ -5502,6 +5502,67 @@ impl<Signer: EcdsaChannelSigner> ChannelMonitorImpl<Signer> {
...
+
+		// Group by payment hash so we fail back all identical HTLCs together
+		let mut htlcs_by_hash: HashMap<PaymentHash, Vec<(&HTLCOutputInCommitment, &HTLCSource)>> = new_hash_map();
+		for (htlc, source_opt) in htlcs {
+			if let Some(source) = source_opt {
+				htlcs_by_hash.entry(htlc.payment_hash).or_default().push((htlc, source));
+			}
+		}
+
+		let monitor_htlc_sources: HashSet<&HTLCSource> = self
```
Comment (TheBlueMatt) [inline lightning/src/chain/channelmonitor.rs]: This logic needs to happen after this matures/reaches 6 confs, not immediately.

**Comment by Anyitechs** [comment]
Pushed new changes to address latest feedback.


EDIT: Working on fixing CI.

**Comment by TheBlueMatt** [comment]
Closing as abandoned. If you're still interested in this, great! Just leave a comment and update it so that CI passes.
